myhpom/tasks.py

```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
 
         return cf_run.pk
 
 
+@shared_task
+class CloudFactoryUpdateDocumentRun(Task):
+    def run(self, cf_run_id):
+        """Update the CloudFactoryDocumentRun object from the CF API.
+
+        ## Use Cases
+        * The admin user can update the status of all or selected runs from CloudFactory.
+
+        ## Parameters
+        * cf_run_id = the id (not the run_id) of the CloudFactoryDocumentRun object.
+        """
+        cf_run = CloudFactoryDocumentRun.objects.get(id=cf_run_id)
+
+        if cf_run.status in [
+            CloudFactoryDocumentRun.STATUS_PROCESSING,  # still going last we checked
+            CloudFactoryDocumentRun.STATUS_TIMEOUT,  # didn't work last time; was it created?
+            CloudFactoryDocumentRun.STATUS_ABORTED,  # we previously aborted; did it take?
+        ]:
+            try:
+                run_url = "%s/runs/%s" % (settings.CLOUDFACTORY_API_URL, cf_run.run_id)
+                response = requests.get(run_url)
+            except requests.exceptions.ConnectTimeout:
+                cf_run.status = CloudFactoryDocumentRun.STATUS_TIMEOUT
+
+            if response.status_code == 404:
+                cf_run.status = CloudFactoryDocumentRun.STATUS_NOTFOUND
+                cf_run.response_content = response.content  # not json, shouldn't raise Exception.
+                cf_run.save()
+            elif response.status_code == 200:
+                cf_run.save_response_content(response.content)
+            else:
+                # don't change the run status here -- we haven't learned anything about it,
+                # and maybe there's a response status that we don't know about
+                raise ValueError(
+                    """URL: %s\nResponse status: %d\nResponse Data: %s"""
+                    % (response.url, response.status_code, response.content)
+                )
+
+
+@shared_task
+class CloudFactoryAbortDocumentRun(Task):
+    def run(self, cf_run_id):
+        """Abort the given CloudFactoryDocumentRun.
+
+        ## Use Cases:
+        * when the user deletes their DocumentUrl, this task is triggered automatically
+            (through the deletion of the DocumentUrl).
+        * can be called from the admin
+
+        ## Parameters:
+        * cf_run_id = the id (not the run_id) of a CloudFactoryDocumentRun object
+
+        ## Process:
+        * If it is in progress (status='Processing'), cancel it at CloudFactory.
+            * 202 = CloudFactory accepted the request
+            * 405 = already processed / aborted at CloudFactory
+        * Get the run info from CloudFactory and update our models.
+            * 200 = Ok
+        * Any response status apart from those given should raise an exception / send error email,
+            because it means we have a bug / false assumption somewhere in our system.
+        """
+        cf_run = CloudFactoryDocumentRun.objects.get(id=cf_run_id)
+        if cf_run.status in [
```

Comment:
AC: "Admin user can verify that a document review was cancelled." The abort task doesn't do that verification – it only stores that CF accepted the abort request. But I am envisioning a case where they accept the abort request but are actually already processing the run and finish processing it. So the result is that the processing was not in fact aborted. The "Aborted" status is non-final – it leaves room for a final verification that the abort was finished and the status="Processed". — Maybe the reality with CF is if they accept the abort request, then we can call the run aborted and be done with it.

When we update the response_content field on 202 or 405, we're not losing anything – the .save_response_content() only updates fields that are given. In fact, that method doesn't do anything in this case except save the status, so I've adjusted the task just to do that. This also removes the possibility of an error on non-json response body, since we don't do anything with the body.

So I don't think we need a new run for the "abort". Since the run model tracks the status of one CF run / process, it makes sense to me to keep it as a single object. In the case of errors, it's always possible to re-submit the document in a different run, but that too is a single run object.

The if-block serves the admin users and the rest of us: If an admin selects all the runs and then hits the 'Abort' admin action, it only actually processes those that ought to be processed – namely, the STATUS_PROCESSING and STATUS_REQ_ERROR (was STATUS_TIMEOUT). I include the second of these because we don't know if the run actually "took" at CF during that request — it's potentially a live run, so this gives us a chance to abort any "potential run". 

We could move the if-block to the admin if that makes more sense to you. I just didn't think the admin action should have to know much about the different STATUS_CHOICES in the model.

Reply:
Yes, I think it's okay – we can treat STATUS_ABORTED as a final status, but if the admin wants to get an update from CF, they can and it becomes STATUS_PROCESSED with 'Aborted' in the response_content.